File: Searching and Sorting Algorithms/Shell Sort/shellSort.cpp
#include <iostream>
#include <vector>

using std::vector;
using type = int;

void printArray(const vector<type>& array) {
  for (const auto& v : array)
    std::cout << v << " ";
  std::cout << "\n";
}

void shellSort(vector<type>& array) {
  int size = array.size();
  int interval = size / 2;
  while (interval > 0) {
    for (int i = interval; i < size; i++) {
      type temp = array[i];
      int j = i;
      while (j >= interval && array[j - interval] > temp) {
        std::swap(array[j], array[j - interval]);
        j -= interval;
      }
      array[j] = temp;
    }
    interval /= 2;
  }
}

int main(int argc, char** argv) {
  vector<type> array{9, 8, 3, 7, 5, 6, 4, 1};
  std::cout << "Unsorted array: ";
  printArray(array);
  shellSort(array);
  std::cout << "Sorted array: ";
  printArray(array);
  return 0;
}

File: Searching and Sorting Algorithms/Shell Sort/shellSort.py
def shellSort(array):
    size = len(array)
    interval = size//2
    while interval > 0:
        for i in range(interval, size):
            temp = array[i]
            j = i
            while j >= interval and array[j-interval] > temp:
                array[j] = array[j-interval]
                j -= interval

            array[j] = temp
        interval //= 2


array = [9, 8, 3, 7, 5, 6, 4, 1]
print("Unsorted Array:", array)
shellSort(array)
print("Sorted Array:", array)
